File: examples-py/write_code.py
import os
import sys
import asyncio
from pydantic import BaseModel
import openai
import tiktoken

# Load environment variables from .env file
from dotenv import load_dotenv
load_dotenv()

# Get code to write
task = ' '.join(sys.argv[1:])
if not task:
    print("Please tell me what code to write.")
    sys.exit(1)

# Initialize OpenAI 
api_key = os.getenv('OPENAI_API_KEY')
model = 'gpt-4o-2024-08-06'


# Define output shape
class SourceCode(BaseModel):
    code: str

# Provide additional context
context = ""

# Ask a question using chain of thought
max_tokens = 4000
goal = f"Write a small typescript program that {task}."

async def generate_object(goal, max_tokens, context):
    try:
        response = openai.Completion.create(
            model=model,
            prompt=goal + "\n" + context,
            max_tokens=max_tokens
        )
        return response.choices[0].text.strip()
    except Exception as e:
        print(f"Error: {e}")
        return None

async def main():
    result = await generate_object(goal, max_tokens, context)
    if result:
        print(f"\033[32m{result}\033[0m")
    else:
        print("Failed to generate code.")

if __name__ == "__main__":
    asyncio.run(main())

File: examples/write-code.ts
import { openai, generateObject, JsonSchema } from "agentm-core";
import * as dotenv from "dotenv";

// Load environment variables from .env file
dotenv.config();

// Get code to write
const task = process.argv[2];
if (!task) {
    console.error("Please tell me what code to write.");
    process.exit(1);
}

// Initialize OpenAI 
const apiKey = process.env.OPENAI_API_KEY!;
const model = 'gpt-4o-2024-08-06';
const completePrompt = openai({ apiKey, model });


// Define output shape
interface SourceCode {
    code: string;
}

const jsonSchema: JsonSchema = {
    name: 'SourceCode',
    schema: {
        type: 'object',
        properties: {
            code: { 
                type: 'string',
                description: 'typescript source code' 
            }
        },
        required: ['code'],
        additionalProperties: false
    },
    strict: true
};

// Provide additional context
const context = ``;

// Ask a question using chain of thought
const maxTokens = 4000;
const goal = `Write a small typescript program that ${task}.`;
generateObject<SourceCode>({goal, jsonSchema, maxTokens, context, completePrompt }).then(result => {;
    if (result.completed) {
        console.log(`\x1b[32m${result.value?.code}\x1b[0m`);
    } else {
        console.error(result.error);
    }
});
